partitura/io/importdcml.py

```diff
@@ -0,0 +1,278 @@
+    data["duration_div"] = duration_div
+    data["alter"] = alter
+    data["pitch"] = data["midi"]
+    grace_mask = ~data["gracenote"].isna().to_numpy() if "gracenote" in data.columns else np.zeros(len(data), dtype=bool)
+    data["id"] = np.arange(len(data))
+    # Rewrite Voices for correct export
+    staffs = data["staff"].unique()
+    re_index_voice_value = 0
+    for staff in staffs:
+        staff_mask = data["staff"] == staff
+        # add re_index_voice_value to the voice values of the staff
+        data.loc[staff_mask, "voice"] += re_index_voice_value
+        # update re_index_voice_value
+        re_index_voice_value = data.loc[staff_mask, "voice"].max()
+
+    note_array = data[["onset_div", "duration_div", "pitch", "step", "alter", "octave", "id", "staff", "voice"]].to_records(index=False)
+    part = spt.Part("P0", "Metadata", quarter_duration=qdivs)
+
+    # Add notes and grace notes
+    for n_idx, note in enumerate(note_array):
+        if grace_mask[n_idx]:
+            # verify that staff and voice are the same for the grace note and the main note
+            note_el = spt.GraceNote(
+                grace_type="grace",
+                id="n-{}".format(note["id"]),
+                step=note["step"],
+                octave=note["octave"],
+                alter=note["alter"],
+                staff=note["staff"],
+                voice=note["voice"],
+                symbolic_duration={"type": "eighth"},
+            )
+
+            if grace_mask[n_idx - 1]:
+                prev_note = note_array[n_idx - 1]
+                for note_prev in part.iter_all(spt.GraceNote, note["onset_div"], note["onset_div"] + 1):
+                    if note_prev.id == "n-{}".format(prev_note["id"]):
+                        note_el.grace_prev = note_prev
+                        note_prev.grace_next = note_el
+                        break
+        else:
+            symbolic_duration = estimate_symbolic_duration(note["duration_div"], qdivs)
+            note_el = spt.Note(
+                    id="n-{}".format(note["id"]),
+                    step=note["step"],
+                    octave=note["octave"],
+                    alter=note["alter"],
+                    staff=note["staff"],
+                    voice=note["voice"],
+                    symbolic_duration=symbolic_duration)
+        part.add(note_el, start=note["onset_div"], end=(note["onset_div"]+note["duration_div"]))
+
+    # Curate grace notes
+    grace_note_idxs = np.where(grace_mask)[0]
+    for i, grace_el in enumerate(part.iter_all(spt.GraceNote)):
+        grace_idx = grace_note_idxs[i]
+        note = note_array[grace_idx]
+        # Find the next note in the same staff and voice
+        if not grace_mask[grace_idx+1]:
+            i = 1
+            next_note = note_array[grace_idx+i]
+            while note["staff"] != next_note["staff"] or note["voice"] != next_note["voice"]:
+                i += 1
+                next_note = note_array[grace_idx+i]
```

Comment:
add a ending condition for this search that will result in ignoring the grace note and print a warning

Reply:
on it